source/encoding.c

```diff
@@ -0,0 +1,316 @@
+    size_t decoded_length = (to_decode_len + 1) >> 1;
+
+    if (!output) {
+        *output_size = decoded_length;
+        return AWS_OP_SUCCESS;
+    }
+
+    if (*output_size < decoded_length) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    size_t written = 0;
+
+    size_t i = 0;
+
+    //if the buffer isn't even, prepend a 0 to the buffer.
+    if (to_decode_len & 0x01) {
+        i = 1;
+        output[written++] = char_to_int(to_decode[i]);
+    }
+
+    for (; i < to_decode_len; i += 2) {
+        uint8_t value = char_to_int(to_decode[i]) << 4;
+        value |= char_to_int(to_decode[i + 1]);
+        output[written++] = value;
+    }
+
+    *output_size = written;
+
+    return AWS_OP_SUCCESS;
+}
+
+static size_t calculate_base64_decoded_length(const char *input, size_t len) {
+    if (len == 0) {
+        return 0;
+    }
+
+    size_t padding = 0;
+
+    if (len >= 2 && input[len - 1] == '=' && input[len - 2] == '=') { /*last two chars are = */
+        padding = 2;
+    }
+    else if (input[len - 1] == '=') { /*last char is = */
+        padding = 1;
+    }
+
+    return (len * 3 / 4 - padding);
+}
+
+static size_t calculate_base64_encoded_length(size_t buffer_len) {
+    return 4 * ((buffer_len + 2) / 3) + 1; /* plus one for the NULL terminator */
+}
+
+int aws_base64_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size) {
+
+    size_t encoded_length = calculate_base64_encoded_length(to_encode_len);
+    if(!output) {
+        *output_size = encoded_length;
+        return AWS_OP_SUCCESS;
+    }
+
+    if(*output_size < encoded_length) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    *output_size = encoded_length;
+
+    size_t buffer_length = to_encode_len;
+    size_t block_count = (buffer_length + 2) / 3;
+    size_t remainder_count = (buffer_length % 3);
+    size_t str_index = 0;
+
+    for(size_t i = 0; i < to_encode_len; i += 3 )
+    {
+        uint32_t block = to_encode[ i ];
+
+        block <<= 8;
+        if (i + 1 < buffer_length)
+        {
+            block = block | to_encode[ i + 1 ];
+        }
+
+        block <<= 8;
+        if (i + 2 < to_encode_len)
+        {
+            block = block | to_encode[ i + 2 ];
+        }
+
+        output[str_index++] = base64_encoding_table[(block >> 18) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 12) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 6) & 0x3F];
+        output[str_index++] = base64_encoding_table[block & 0x3F];
+    }
+
+    if(remainder_count > 0)
+    {
+        output[block_count * 4 - 1] = '=';
+        if(remainder_count == 1)
+        {
+            output[block_count * 4 - 2] = '=';
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static void init_decoding_table() {
+    if(!base64_decoding_table_init) {
+        for(size_t i = 0; i < (sizeof(base64_encoding_table) / sizeof(char)); ++i)
```

Comment:
Let's make this a compile-time-constant table. It's a minor pain but it'll make the actual code at runtime simpler and safer.